neusomatic/python/extend_features.py

```diff
@@ -0,0 +1,378 @@
+import multiprocessing
+import os
+import gzip
+
+import pysam
+import numpy as np
+
+import sequencing_features
+import genomic_file_handlers as genome
+from read_info_extractor import rescale
+
+
+def extract_features(candidate_record):
+    reference, tumor_bam, normal_bam, min_mapq, min_bq, dbsnp, batch = candidate_record
+    thread_logger = logging.getLogger(
+        "{} ({})".format(extract_features.__name__, multiprocessing.current_process().name))
+    try:
+        tbam = pysam.AlignmentFile(tumor_bam)
+        nbam = pysam.AlignmentFile(normal_bam)
+        ref_fa = pysam.FastaFile(reference)
+        if dbsnp:
+            dbsnp_tb = pysam.TabixFile(dbsnp)
+
+        ext_features = []
+        for chrom, pos, ref, alt, if_cosmic, num_cosmic_cases in batch:
+            var_id = "-".join([chrom, pos, ref, alt])
+            pos = int(pos)
+            my_coordinate = [chrom, pos]
+            nBamFeatures = sequencing_features.from_bam(
+                nbam, my_coordinate, ref, alt, min_mapq, min_bq)
+            tBamFeatures = sequencing_features.from_bam(
+                tbam, my_coordinate, ref, alt, min_mapq, min_bq)
+
+            n_ref = nBamFeatures['ref_for'] + nBamFeatures['ref_rev']
+            n_alt = nBamFeatures['alt_for'] + nBamFeatures['alt_rev']
+            t_ref = tBamFeatures['ref_for'] + tBamFeatures['ref_rev']
+            t_alt = tBamFeatures['alt_for'] + tBamFeatures['alt_rev']
+            sor = sequencing_features.somaticOddRatio(
+                n_ref, n_alt, t_ref, t_alt)
+
+            homopolymer_length, site_homopolymer_length = sequencing_features.from_genome_reference(
+                ref_fa, my_coordinate, ref, alt)
+
+            indel_length = len(alt) - len(ref)
+
+            if_dbsnp = 0
+            if_common = 0
+            if dbsnp:
+                region = "{}:{}-{}".format(chrom, pos, pos + 1)
+                dbsnp_vars = {}
+                for x in dbsnp_tb.fetch(region=region):
+                    chrom_, pos_, _, ref_, alts_, _, _, info_ = x.strip().split("\t")[
+                        0:8]
+                    for alt_ in alts_.split(","):
+                        dbsnp_var_id = "-".join([chrom_, pos_, ref_, alt_])
+                        dbsnp_vars[
+                            dbsnp_var_id] = 1 if "COMMON=1" in info_ else 0
+                if var_id in dbsnp_vars:
+                    if_dbsnp = 1
+                    if_common = dbsnp_vars[var_id]
+
+            p_scale = None
+            CHROM = my_coordinate[0]
+            POS = my_coordinate[1]
+            REF = ref
+            ALT = alt
+            if_dbsnp = if_dbsnp
+            COMMON = if_common
+            if_COSMIC = if_cosmic
+            COSMIC_CNT = num_cosmic_cases
+            Consistent_Mates = tBamFeatures['consistent_mates']
+            Inconsistent_Mates = tBamFeatures['inconsistent_mates']
+            N_DP = nBamFeatures['dp']
+            nBAM_REF_MQ = '%g' % nBamFeatures['ref_mq']
```

Comment:
Can this long list of entry names be looped? Seems like the logic is simple.

Reply:
I tried to keep SomaticSeq style. For now, I would rather leave it like this.